### ISSUE TITLE
Add the ability for AutoRoutePart to have additional aliases for a content item

### ISSUE THREAD

**Issue by mahsaro** [OrchardCMS/Orchard#5617]
### Purpose ###
1. Content item to be accessed through multiple URLs. 
2. Ensuring SEO compatibility by use of canonical URLs.

### Requirements ###
1. Admin must be able to create multiple alias paths for a Content Item. 
2. Admin must be able to select only one path as the preferred URL(Primary Path) for the Content Item.
3. Admin must be able to either **create a path manually** or **select a pattern** to generate the path.
4. Admin must be able to edit Content Item's paths, by altering the pattern or modifying the text manually. The regeneration of paths must take place after Content Item is saved.
5. Change of patterns in setting must **not** affect the existing alias paths.
6. If the regenerate path option is selected in the settings, paths must be regenerated when Content Item is modified and saved. 
7. The preferred URL will be indicated as canonical. 

### UI Interactions ###
Following changes will take place in Content Item editor including an Autoroute part:

1. Admin will be able to add multiple paths instead of only one that is currently available.
2. "Permalink" will be renamed to "Path".
3.  Admin can select one of the paths (by checking a radio button) as the primary path which would be used as the preferred URL to access the Content Item.
4. When creating/editing a Content Item, Admin will be able to choose one of these options from a dropdown list for each path:
    *  "Custom Path" which enables admin to provide a path manually. This option **enables** the textbox for indicating the path manually.
    *  A pattern from a list of patterns retrieved from the part's settings. This option **disables** the textbox for indicating the path manually. If nothing is chosen, alias will be made based on the default pattern selected in Autoroute Part Settings.
5. When editing a Content Item, for existing paths generated by patterns, Admin will see the selected pattern in a dropdown list and the generated path in a disabled textbox. If a new pattern is selected, the disabled textbox will be emptied. The newly generated path will be shown once the Content Item is saved.
6. When editing a Content Item, for existing custom paths, Admin will see "Custom Path" selected in the dropdown list and the manually created path in a enabled textbox.

### Implemetation ###
1. AliasPathInfo.cs to be created with following properties:
   * string AliasPath
   * int? PatternIndex (if null, that mean that the path is made manually)
2. AutoroutePart.cs will be modified to contain the following properties:
   * IEnumerable<AliasPathInfo> Paths
   * int PrimaryAliasPathIndex (index of the preferred path used as the canonical URL)

**Comment by dcinzona** [comment]
I like this as it will remove the need to manage custom aliases in a different module... although, the aliases module is getting updated to support better organization of aliases (those auto-generated vs user-generated).  It would still be nice to be able to manage these aliases directly via the aliases UI admin page.  Not sure if there will need to be any additional updates to that as well.

**Comment by sfmskywalker** [comment]
Not sure if you would need to be able to manage these AutoroutePart-managed aliases fro, the Alias UI, since you can always just click on the path from that UI which will take you directly to the content item, from where you can add/remove/edit the various aliases anyway. The managed aliases are managed by the system, so allowing the user to manage them from the Alias UI potentially complicates things and may be confusing even.

In any case, really like the idea of the ability to create multiple aliases per content item in an easy manner.

**Comment by DanielStolt** [comment]
@dcinzona Another point on this: all this actually came out of the realization that custom aliases can't reliably be used to accomplish this, as they will store the ID values of the items to which they point, and those ID values are not persistent across export/import.

**Comment by dcinzona** [comment]
True, but I think we should still have some kind of view where an admin can list out all of the site aliases and easily get to the content item edit page that contains it.  Just an easier way to figure out what content is where and how users are getting to it.  
There are some issues with the custom aliases module that we are using anyway, so we would like to get away from that...but in the cases where we would like to create short links (or convenience links) to the same content item / url, that's what we have to use for now.

**Comment by dcinzona** [comment]
@sfmskywalker I think what I was trying to say was that we would have to make it so we couldn't manage aliases from the Alias UI - or when adding an alias, instead of providing you with a text field for a route path, it should be a content item picker.

**Comment by sebastienros** [comment]
Why not use permanent redirects instead? You would have a canonical url for a content item, and then the redirects would point to this. So when we render the content item url, it will always be the same. This could still be implemented as a separate module or feature, even with a custom content part to set alternative urls. And there could also be another main page to manage all redirects. You can even use it for controller/action.

**Comment by DanielStolt** [comment]
@dcinzona You would still be able to view these additional aliases in the aliases UI, as they would be listed under the "managed" tab. Just like other autoroute aliases, they will appear there but in read-only form (you won't be able to edit/delete them). Any modifications to them, you are expected to perform from the autoroute part UI.

@sebastienros Why not have the option of both? When you create an additional path for a content item, you can specify if it should act like a redirect or as an alternate alias?

**Comment by mjy78** [comment]
Can this also support redirecting from the default content display url (/Contents/Item/Display/10) to the autoroute url?

**Comment by dcinzona** [comment]
@mjy78 That's an interesting option... Maybe as a setting to enable it globally or within each content item individually? There could be scenarios where you would want it to redirect to the autoroute url, and others where you wouldn't. (i.e. displaying widgets / layers)
